Implemented LightingEngine to improve multi light change handling

Mathe172 opened the pull request:
This PR is the third one in our attempt to fix and optimize the vanilla lighting engine. For more details see #3848 (the stuff done here is actually the main intent of the linked issue).

This particular PR has the following effects:
* Fixes [MC-102162](https://bugs.mojang.com/browse/MC-102162)
* Drastically improves the performance of multi-block changes like /fill commands

A quick overview of what it does:
* Processing light updates is postponed until the result is actually needed.
* We completely rewrote the lighting engine (following Vanilla logic). We think that replacing this methods completely shouldn't impose any compatibility issues, as the main logic behind the lighting model should never change.

To be more precise: (some additional comment on the implementation can be found in the code)
* Each `World` now has an instance of `LightingEngine`
* Calls to `World.checkLight` are redirected to `LightingEngine.scheduleLightUpdate`
* Light updates are actually processed at the following places:
  * `Chunk.getLightFor` and `.getLightForSubtracted`. This covers most accesses to the light values.
  * `AnvilChunkLoader.saveChunk`, to ensure that updates are not lost upon quitting the game
  * `SpacketChunkData`, to ensure that they are correctly synced to the client
  * `Minecraft.runTick`, to ensure that lighting updates are processed at least once a tick (before lighting notifications are processed)
* Processing the light updates sorts them by whether the block should get darker or brighter and filters out duplicate schedulings of the same block
* Brightenings and darkenings are processed simultaneously from brightest to darkest. This ensures a good order of updates, so that a single pass suffices and each block is typically updated only once or twice (because it has to be brightened again after darkening).
* The main logic behind this follows the Vanilla code. We just reordered the updates for better performance for multi-block updates.

Reasons why we didn't patch the Vanilla method:
* We need `long`s to store the update positions, while Vanilla uses ìnt`s (which assumes that updates are confined in a 32x32x32 area). Also the Vanilla queue has a fixed size of 32'768 which is too small in general. Patching this would require to change all the bit operations
* We reoder the order in which updates are processed to achieve better performance for multi-block changes. This can't really be patched in with a few lines.
* We cache chunk lookups where possible, as these are taking up ~30% of the exection time for the Vanilla engine.

Remarks:
* Client-side, we explicitly ensure that we are on the main thread. This is because the rendering threads access the world unsynchronized, so they shouldn't cause any modifications
* We bypassed the call to `Chunk.propagateSkylightOcclusion` as the check causes he light updates to be processed immediately, which negates the purpose of this PR. This patch is done properly in #3871, so that one should be merged first (in case this should be accepted).
* There are still some lighting bugs which we plan to fix by other PRs
* To check that we cover all places where we need to proc updates, simply search for accesses to `ExtendedBlockStorage.blockLightArray` and `.skylightArray`, as well as for implementations of `IWorldEventListener.notifyLightSet`
* We implemented a `PooledLongQueue` since the queues change sizes a lot. Not trimming them would result in a large amount of idle memory. On the other hand, trimming them to a small size would lead to frequent reallocations. As the queues are not filled up simultaneously, but the entries move as the algorithm proceeds, the `PooledLongQueue` effectively reduces the overall memory footprint by ~1/15 (compared to former alternative), while reducing allocations to (nearly) zero
* We change one thing w.r.t. to Vanilla: While Vanilla ensures that the neighbor chunks are loaded before light updates are done (which may create defects inside a chunk), we propagate them as far as possible (which moves any such defect to the chunk boundary). This has two upsides:
  * It might be the case that the light changes don't even propagate that far
  * If one tries to fix these glitches later (done in #4184), you only need to search the borders instead of the whole chunk
* World-gen performance (as well as single block changes) is not affected
  * What makes world-gen so expensive is `Chunk.checkLight` (~30% of the total world-gen time), which is actually spent checking the (mostly) dark underground. We may target this in a future PR building up on this one.
* Due to some of the optimizations, the new LightingEngine is less likely to fix lighting glitches "on-the-fly" (i.e. light defect in a region that is updated by some other change). However, this is not an issue if we fix all problems. Additionally, the fixing done by the Vanilla code is spotty at best.
* [EDIT] This PR also depends on #3870, since Vanilla does not initialize sections in `Chunk.setBlockState` when a block exists above them, which is fine if light updates are carried out immediately (because `Chunk.setLightFor` would have initialized them already if they were non-trivial), but not in our case (where `Chunk.setLightFor` is not immediately called). #3870 causes the sections to be initialized in any case, which fixes any possible issues[/EDIT]

Commit:
Implemented LightingEngine to improve multi light change handling

Added LightingEngine & added instance to World
Replaced logic in World.checkLight to only schedule light updates
Chunk.getLightFor processes queued updates to ensure correct results
Added Chunk.getCachedLightFor that suppresses update processing
Added calls to LightingEngine.procLightUpdates at following places:
* Chunk.getLightFor
* Chunk.getLightForSubtracted
* AnvilChunkLoader.saveChunk (accesses raw light arrays)
* SPacketChunkData constructor (accesses raw light arrays)
* Minecraft.runTick (before light notifications are processed)
LightingEngine.procLightUpdates does nothing on client non-main threads

covers1624 commented on src/main/java/net/minecraftforge/common/lighting/LightingEngine.java:
```diff
@@ -0,0 +1,630 @@
+        }
+
+        //renderer accesses world unsynchronized, don't modify anything in that case
+        if (this.world.isRemote && !Minecraft.getMinecraft().isCallingFromMinecraftThread())
```
Should this check be done first?

Mathe172 replied:
Since this check is the more expensive of the two, it's better to do it later: The only thing that could happen is that the render-thread (or similar) wrongfully returns on the first check due to some race condition - which is exactly what we wanted anyway, it just happened without doing this more expensive check

fewizz replied:
Isn't 'Minecraft' is a client-side class? Do you tested this on server?

williewillus replied:
yup, this will crash on dedicated server

Mathe172 replied:
Thanks for pointing this out. While it will not crash when using the Oracle JVM (as the method is never actually called on a dedicated server), [it is apparently not specified](http://stackoverflow.com/a/4744634/3032376) (the Oracle JVM does the loading lazily, which is why it only throws exceptions when the class is actually needed). We'll address this in the next commit.
There are however at least two similar issues already in Forge: (should we make an issue/PR for these?)
* `ForgeHooks.onPickBlock` references `Minecraft` - this one should probably be in `ForgeHooksClient`
* `FMLCommonHandler` references `FMLServerHandler.instance()` (that's actually the whole line), which in turn references `DedicatedServer`.

LexManos replied:
Referencing client side code is fine, you just have to be sure not to load the class/method thats references it.
Or reflectively loop through the class that references it. ForgeHooks is fine because there should be no reason for people to reflect through. In this case however ass lighting is done both sides you'd probably want to wrap this around a FMLHandler

Mathe172 replied:
I'm not sure I understand what you mean: Isn't the situation the same for `ForgeHooks` (which references `Minecraft` in `onPickBlock(...)`, line 496) and `LightingEngine` (the only difference I see is that the method containing the call is not referenced from server code in the case of `ForgeHooks.onPickBlock`)? In both cases, the offending code will never be executed, but could cause problems (i.e. a `ClassNotFoundException` might be thrown) depending on how resolution is implemented in the JVM. The fact that the method is not reference doesn't seem to matter [according to the JLS](https://docs.oracle.com/javase/specs/jls/se6/jls3.pdf) (page 310, last paragraph).

LexManos replied:
If the METHOD is resolved. Then all classes are linked. So the current method would get a class not found exception if it is either called, or it's reflected over.

Take the following for example:
```
class Test {
  public void someMethod() {
    if (!remote && ClientClass.func()) {}
  }
}
```
If someone CALLS that, or Does Test.class.getDeclaredMethods() then it will throw a ClassNotFoundException

Now take this:
```
class Test {
  public void someMethod() {
    if (!remote && clientBouncer()) {}
  }
  public boolean clientBouncer(){
    return ClientClass.func();
  }
}
```
Now, if someone calls `someMethod` then it will NOT throw an error, as long as remote is true.
That's because the JVM is lazy-linked {In all JVM implementations i've seen}. So when it compiles the 'someMethod' function is only cares that it's calling clientBouncer and it lazy compiles so it will only compile clientBouncer if the logic ever causes the function to be called.
Now, Reflection throws a wrench in the works. Getting the Method/MethodHandle for a method causes the JVM to compile that method, so doing Test.class.getDeclaredMethods would throw an error because it's trying to compile clientBouncer and running into a client only method.

Get it?

As for ForgeHooks, that class is not reflected over, and the only place onPickBlock is called is from specifically the Minecraft class which is purely client code, so we can get away with it. If some other mod comes along and reflects all the methods in ForgeHooks they will crash, and THEY will be responsible for it. 

As for what the JLS states, there is a MAJOR difference between what it states and what actually exists in the real world. I deal in reality not specs. However, if you REALLY want to be safe, like I said, you could use the FML*Handler classes, this type of case is exactly what they are designed for.

PhiProven replied:
We already did that :)

Thanks for the explanation
However, I've tested the old code on a dedicated server (I've built the installer and executed it, instead of running the server with IntelliJ) and it worked. So, apparently the reference to `Minecraft.getMinecraft()` didn't cause it to be resolved.
Anyway, the code should be safe now :)

williewillus commented on src/main/java/net/minecraftforge/common/lighting/LightingEngine.java:
```diff
@@ -0,0 +1,630 @@
+            ? this.curChunk.canSeeSky(this.curPos)
+            ? EnumSkyBlock.SKY.defaultLightValue
+            : 0
+            : MathHelper.clamp(state.getLightValue(this.world, this.curPos), 0, MAX_LIGHT);
```
nested ternaries are harder to mentally parse than if statements and they take about the same lines :P

Mathe172 replied:
We'll change it for the next commit

williewillus commented on src/main/java/net/minecraftforge/common/lighting/LightingEngine.java:
```diff
@@ -0,0 +1,630 @@
+    private final Chunk[] neighborsChunk = new Chunk[EnumFacing.VALUES.length];
+    private final MutableBlockPos[] neighborsPos = new MutableBlockPos[EnumFacing.VALUES.length];
+    private final int[] neighborsLight = new int[EnumFacing.VALUES.length];
+
```
how many of these *must* be up here?

I'd rather pass around more parameters than have more state variables, but it's just my preference :P

Mathe172 replied:
We could remove `lightType`, if that's preferred... 
Normally, we also prefer parameters, but for these, we decided to make an exception for the following reasons:
* We need the neighbor caching data, unless we pass some caching wrapper object back and forth, which is even uglier. The other alternative would be to inline `fetchNeighborData`
* The iteration state variables are needed for a similar reason (since we'd need to inline `nextItem` at 5 places)

Commit:
Removed reference to Minecraft in LightingEngine

Added FMLCommonHandler.getMinecraftThread()

Commit:
Replaced nested ternary with 'if' for readability

Mathe172 commented:
We failed to mention that this PR also needs #3870 to work properly in all edge cases - we updated the description accordingly

Commit:
Fixed possible glitch caused by asynchronous access by rendering threads

Commit:
Optimized lighting engine to reduce number of light notifications

Reordered brightening & darkening
Do not respread inside darkening immediately, schedule them instead
Shortcut darkening when already brightened

Commit:
Changed position shifting to long arithmetic

Reordered long layout for position to enable overflow detection on y
Added neighborShifts for faster neighbor position calculation

Commit:
Reuse current chunk if possible in nextItem() and fetchNeighborData()

Commit:
Removed unnecessary call to posToChunk in favor of cache lookup

Commit:
Move call to LightingEngine.procLightUpdates() before chunk unloading

PhiProven commented:
To give a benchmark:
You can replace some area with `stone` and `sea_lantern` alternately with a 10 Hz clock.

Vanilla already lags quite a bit for a `5x5` area. Using a `50x50` area needs several seconds for a single tick.
This PR can still handle the `50x50` area smoothly.

You can find the test world [here](https://github.com/MinecraftForge/MinecraftForge/files/1293644/Lighting.Test.zip)

Commit:
Use indexed loops

mrgrim commented:
While testing these patches I noticed that the nether lava sea doesn't get lit properly. It seems blocks with a light value are never checked when built via ChunkPrimer. Vanilla kind of solves this by calling the Chunk method "resetRelightChecks()" when Nether chunks are generated and letting the "enqueueRelightChecks()" get ticked via "updateBlocks()" in the WorldServer class.

This isn't an issue in the overworld because all light sources are part of the population stage of world gen.

In vanilla the first light check on the surface of a lava lake by "enqueueRelightChecks()" combined with the heavy handed approach to lighting in the vanilla "checkLightFor" functions seems to take care of getting things lit "good enough" while the full chunk relight ticks away. However, with these patches you can see in real time as column by column the lakes start to light up, and it seems easy to accidentally break the process.

I think perhaps performing light checks for blocks with a light value in the Chunk constructor that takes a ChunkPrimer object would take care of this.

PhiProven commented:
I was quite sure that I checked the lighting in the nether....
The checking you described is done by `Chunk.checkLight()`. For whatever reason, it is only carried out if the world has skylight, though it mostly checks for block light sources.

PhiProven commented:
Removing the `this.world.provider.hasSkyLight()` check inside `Chunk.checkLight()` indeed seems to be sufficient to solve the issue. Alternatively, we could explicitly check for `WorldProviderHell`, though I don't really like that.
Enabling the check will certainly have a negative impact on performance, but I don't think that it's really that much. IIRC `Chunk.checkLight()` takes up ~2% of worldgen time (only taking into account the time spent by the method itself, thus excluding the lighting engine).
What do you think about enabling `Chunk.checkLight()` for all worlds, regardless of whether the world has skylight? If there are modded worlds that don't have skylight and place light sources via ChunkPrimer they should have the same issue already with Vanilla/Forge (It is not really visible for lava lakes because the first run of `enqueueRelightCheck()` almost always hits a light source).

mrgrim commented:
It works for the Nether, but I'm not sure it's the best approach. It's certainly the least invasive. Keep in mind I'm just some bloke that found your patches and decided to try them out.

There are a couple of issues I think with relying on checkLight. First, its logic is very much skylight oriented. For example, it doesn't issues a "checkLight" call on any block with a non 0 opacity below either sea level or, if there are no opaque blocks above sea level, the below the first solid block below sea level. This combines with the second issue.

The other thing is that when checkLight is called ExtendedBlockStorage still has the light level set to 0 for all blocks generated with ChunkPrimer, even blocks with a light value set. setBlockState in the World class sets this value for all blocks in the population phase, but blocks from Primer will be inconsistent, which your lighting engine doesn't plan for in "fetchNeighborDataFromCur". This, for example, would result in no light being calculated when an air block next to an opaque light emitting block above sea level is checked because the opaque block is not itself checked due to above.

It seems to me this really needs to be done in the Chunk constructor when it's converting ChunkPrimer's array of char's into proper block states. It seems like ExtendedBlockStorage should be initialized and checkLightFor of the block done there to really cover chunk initialization properly.

However, I leave it up to you and the Forge devs, as these patches are already quite considerable.

PhiProven commented:
I'll work on that tomorrow. I think it makes sense to put this in a separate PR, because it should probably also remove the black spots that are experienced in the Nether, which is a Vanilla bug (The lighting above the lava lake will still be wrong due to the 16 block spreading limitation of Vanilla but it should look better, at least).
Thanks for pointing out the issue @mrgrim

mrgrim commented:
I've been getting infrequent crashes while rendering entities from inside of procLightUpdates. It seems something somewhere isn't thread safe. I have a couple of crash reports. Is this an appropriate venue? My email address is on my profile.

PhiProven commented:
That's interesting. `if (this.world.isRemote && FMLCommonHandler.instance().getMinecraftThread().isCallingFromMinecraftThread())` should take care of concurrency...
Can you link the crash reports here?

mrgrim commented:
Sure. I saw that and tried to verify it was working as intended, but I'm not at all familiar with how threading is done in MC. Reports should be attached.
[crash-2017-09-24_21.58.30-client.txt](https://github.com/MinecraftForge/MinecraftForge/files/1329597/crash-2017-09-24_21.58.30-client.txt)
[crash-2017-09-25_01.04.28-client.txt](https://github.com/MinecraftForge/MinecraftForge/files/1329598/crash-2017-09-25_01.04.28-client.txt)

One thing I should add is that I implemented the method of lighting Chunk's generated with a ChunkPrimer. It runs World.checkLightFor(EnumSkyBlock.BLOCK, ...) on all blocks with a light value > 0 in the Chunk classes "onLoad" method, but only if the chunk was constructed with a ChunkPrimer. I suspect this has added hundreds of additional lighting calls on chunk generation and impacted the distribution of when lighting updates are waiting in the queue.

I suspect this just made the crash more likely and didn't cause it, but I can't be sure.

PhiProven commented:
Unfortunately, I can't reproduce the bug. Can you provide a little more information?
* Which commit did you use exactly?
* Did you use any non standard settings?
* What did you do to provoke the bug? Just fly around in the Nether?
* How long did you have two wait for the bug to occur?

Thanks for your help :-)

mrgrim commented:
I just realized I'm in the 1.11.x version of this PR, but I'm actually doing this in 1.12. I've implemented all 4 of your 1.12 PR's, multi lighting, lighting cleanup, lighting bugfix, and lighting boundary fix. I also applied your lighting notifications change. All of this is in the "1.12.x-vanilla-fixes" branch of my fork of Forge here in github including the Chunk loading modifications if you want to see it.

Only settings changes are unlimited framerate and no vsync. I'm usually pushing 300-500fps. I alway sload a fresh creative world, immediately build a nether portal, then I fly around. I noticed the crashes tended to be when I stopped to smell the roses and moved around unpredictably in a smaller area. I've only had the two so far.

Commit:
Reset chunk cache for every run of the LightingEngine

PhiProven commented:
The only explanation I can think of is the following:
We suppressed the null check for `curChunk` in line 243 of the `LightingEngine` because the chunk was nonnull when the pos was added to `initialBrightenings` and the chunk couldn't be unloaded in the meantime (as no concurrency is involved in unloading).
Unfortunately, we forgot to reset the chunk cache between runs of the lighting engine. Hence we may have cached a chunk that was unloaded between the runs of the lighting engine, and hence wrongfully considered it loaded. When the code reached line 243, the chunk was looked up again, but couldn't be found (because it was unloaded).
Hopefully, the last commit should fix the issue.

mrgrim commented:
I can't believe I missed that! I just flew around for about 10 minutes over a few thousands blocks weaving about and performing sudden reversals of direction, and so far so good. If there's another crash, I'll find it though. A nether with perfect lighting is something I could get use to and fly around all day!

Regarding the fix, that seems to expose a problem. If lighting updates cannot complete before unload then, with these patches reliance on correct generation over on the fly correction, that could be a source of glitches, no? Would a call to procLightUpdates on chunk unload be warranted?

PhiProven commented:
`procLightUpdates()` is called before the chunk is saved to disk on the server. Errors that arise due to unloaded neighbor chunks are corrected by #4184.
(Your merge of #4184 is a bit wrong btw. You added the call to `scheduleRelightChecksForChunkBoundaries()` to `onUnload()` instead of `onLoad()` ;-) )

mrgrim commented:
Thanks! I was wondering about that... that could explain the single obvious glitch I encountered that was corrected on relog while trying to induce a crash. I'm definitely keeping these patches in my personal build, and I'll report any other issues I run into. Good luck with getting them approved!
